keps/sig-windows/3746-windows-root-fs-volume-size/README.md

```diff
@@ -0,0 +1,884 @@
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+Allow `resources.limits.storage` to be specified on Containers in pod specs (only if Pod OS is set ot Windows??) and pass the value to container runtime `CreateContainer` calls by setting the existing `WindowsContainerResources.rootfs-size-in-bytes` field in the CRI-API.
+
+Ex:
+
+```yaml
+kind: Pod
+metadata:
+  name: some-pod
+spec:
+  containers:
+  - name: container1
+    resources:
+      limits:
+        storage: 30Gi
+        ephemeral-storage: 40Gi
+        cpu: 500m
+  - name: container2
+    resources:
+      limits:
+        storage: 40Gi
+        cpu: 500m
+  os:
+    name: windows
+```
+
+### User Stories (Optional)
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### Story 1
+
+A user wants to run a workload that requires more than 20Gb of disk space on the root-fs volume.
+
+#### Story 2
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+IO performance when writing to `emptyDir` volumes is worse then writing to container root fs.
+https://github.com/microsoft/Windows-Containers/issues/345
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+There is little risk to this enhancement and CRI-API and container runtimes already have the ability to specify the root-fs volume size for Windows containers, Kubernetes just needs to expose the functionality in the Pod spec.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### API server updates
+
+Update API server validation to allow `resources.limits.storage` to be specified on containers. While the feature is in development this validation will be gated on by a feature flag.
+The relevant code is in [ValidateContainerResourceName](https://github.com/kubernetes/kubernetes/blob/ad18954259eae3db51bac2274ed4ca7304b923c4/pkg/apis/core/v1/validation/validation.go#L73-L86)
```

Comment:
Relaxing validation has to be done over two releases. Otherwise, we allow data to be persisted that will be considered invalid by the previous release of the API server (which could still be running in a multi-server cluster during rolling upgrade).

I would suggest doing the following:

1. in `pkg/apis/core/validation/validation.go`, add an `AllowContainerStorageResource bool` field to `PodValidationOptions`
2. make sure that options struct is passed through all validation functions down to ValidateContainerResourceName
    * this will reveal *lots* of call paths that lead to this function, all of which have to consider whether to allow the relaxed data or not
    * in ValidateContainerResourceName, only allow the relaxed validation if the `AllowContainerStorageResource` field is true
3. in `GetValidationOptionsFromPodSpecAndMeta`, compute the validation options based on the new and old object
    * in this release
      * on create (new object is non-nil, old object is nil), set the `AllowContainerStorageResource` field to `true` only if the feature gate is enabled
      * on update, set the `AllowContainerStorageResource` field to `true` *only* if the old object already has a container using the storage resource or the feature is enabled
    * in a future release, enable the feature gate
    * in an even more future release, graduate the feature gate and remove the option check